modules/dxp/apps/commerce/commerce-shop-by-diagram/commerce-shop-by-diagram-service/src/main/java/com/liferay/commerce/shop/by/diagram/service/impl/CPDefinitionDiagramPinServiceImpl.java

```diff
@@ -43,10 +42,96 @@
 public class CPDefinitionDiagramPinServiceImpl
 	extends CPDefinitionDiagramPinServiceBaseImpl {
 
-	/**
-	 * NOTE FOR DEVELOPERS:
-	 *
-	 * Never reference this class directly. Always use <code>com.liferay.commerce.shop.by.diagram.service.CPDefinitionDiagramPinServiceUtil</code> to access the cp definition diagram pin remote service.
-	 */
+	@Override
+	public CPDefinitionDiagramPin addCPDefinitionDiagramPin(
+			long userId, long cpDefinitionId, int number, double positionX,
+			double positionY)
+		throws PortalException {
+
+		_cpDefinitionModelResourcePermission.check(
+			getPermissionChecker(), cpDefinitionId, ActionKeys.UPDATE);
+
+		return cpDefinitionDiagramPinLocalService.addCPDefinitionDiagramPin(
+			userId, cpDefinitionId, number, positionX, positionY);
+	}
+
+	@Override
+	public void deleteCPDefinitionDiagramPin(long cpDefinitionDiagramPinId)
+		throws PortalException {
+
+		CPDefinitionDiagramPin cpDefinitionDiagramPin =
+			cpDefinitionDiagramPinLocalService.getCPDefinitionDiagramPin(
+				cpDefinitionDiagramPinId);
+
+		_cpDefinitionModelResourcePermission.check(
+			getPermissionChecker(), cpDefinitionDiagramPin.getCPDefinitionId(),
+			ActionKeys.UPDATE);
+
+		cpDefinitionDiagramPinLocalService.deleteCPDefinitionDiagramPin(
+			cpDefinitionDiagramPin);
+	}
+
+	@Override
+	public CPDefinitionDiagramPin getCPDefinitionDiagramPin(
+			long cpDefinitionDiagramPinId)
+		throws PortalException {
+
+		CPDefinitionDiagramPin cpDefinitionDiagramPin =
+			cpDefinitionDiagramPinLocalService.getCPDefinitionDiagramPin(
+				cpDefinitionDiagramPinId);
+
+		_cpDefinitionModelResourcePermission.check(
+			getPermissionChecker(), cpDefinitionDiagramPin.getCPDefinitionId(),
+			ActionKeys.UPDATE);
```

Comment:
This and the following methods will be used in the edit screen of the product. As per requirements, we should use UPDATE key

Reply:
Hey @Commando18 your intention may be not clear to any other developer that may use that get method (for example I may go with some custom code implementation and it may end up that users of my feature can not view data even my view has no intention to update data). You should check UPDATE permission in the moment you submit changes to backend.
I think you want to mirror DBMS features like read-for-update which we have no adequate pattern for.
I'll let this go but I'm pretty sure it would be rejected if Brian notice it.